docs/rfc/0000-radicle_daemon.md

```diff
@@ -0,0 +1,303 @@
+of churn when major changes were made in the latter, causing weeks/months of
+integration work to catch up to the latest and greatest. As well as this, it
+made it harder to gauge whether code being added to `coco` could have been
+useful to be in `librad` instead.
+
+This made us make the first move to migrating the `coco` crate over to
+`radicle-link` under the name `daemon` – see
+[#576](https://github.com/radicle-dev/radicle-link/pull/576).
+
+This RFC wants to tackle the next phase of this plan and give a concrete plan
+for implementing a general purpose `daemon` that can serve `radicle-upstream`
+and any other applications that would benefit from a high-level API on top of
+`librad` et al.
+
+## Overview
+
+To learn from our past mistakes, we would like to move forward in a way that
+identifies core components and design them in such a way that allows us to
+compose them easily, allowing upstream consumers to mix-and-match them in any
+way they desire. This desire leads us to the following architectural layout. A
+core API that defines the capabilities for working with `radicle-link` data. Two
+consumer-level packages HTTP and CLI for building interesting applications and
+workflows from the core. A reactor to our core that defines daemon-level
+protocols and is ultimately the running process for the `daemon`. And finally, a
+git server that is specific to the git implementation of `radicle-link` allowing
+us to use the `git` CLI for Radicle purposes.
+
+## Core
+
+The goal for the `core` is to make it as reusable as possible, all the while
+making sure that it remains stable as an API, only making additions to it rather
+than changing its surface area.
+
+To do this we propose that the core API consist of capabilities that are defined
+as traits. A capability would be defined for a single resource, however, it
+could be split into sub-capabilities, for example, if there is a set of
+read methods and a set of write methods.
+
+Remark: a capability in this case is set of methods that define the ways one can
+interact with the data the capability is defined for. For example, a capability
+for a directory might be `touch`, `ls`, `mkdir`. It is defined as trait and can
+be given multiple implementations.
+
+The following capabilities our already in scope at the time of writing this RFC,
+but naturally, more will be added as the project evolves.
+
+* Gossip
+* Identity
+  * read
+  * write
+* Profile
+* Replication
+* Tracking
+
+We will sketch these capabilities here, but this may not reflect the final
+definition that will be found in the implementation.
+
+### Identity
+
+The identity capabilities define the create, read, and update methods
+for interacting with Radicle's family of `Identity` types. See
+[spec/identities][id] for more details.
+
+```rust
+pub trait read::Identity<I> {
+	type Error;
+	
+	fn get<R>(&self, urn: Urn<R>) -> Result<Option<I>, Self::Error>;
```

Comment:
How will the replication state be represented? From the description below I understand that `I` represents the data of the identity. But we also need information about the replication state. As I understand it the state indicates from which peer we would like data and from which peer we already got the data.

Reply:
One could build the replication state by using the `Rad` capability, i.e. `Rad::delegates` + `Rad::signed` give you maintainers + contributors, and neither means a tracker.

I'm thinking now that I might not list _all_ capabilities but rather provide the verticals and give a taste of what's included in those verticals, and perhaps how they may be combined.